src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_PrimaryWorkspace_CodeLens.cs

```diff
@@ -0,0 +1,168 @@
+                        syntaxNode, token).ConfigureAwait(false);
+                }
+            }, cancellationToken);
+        }
+
+        public void SetCodeLensReferenceCallback(Guid projectIdGuid, string filePath, CancellationToken cancellationToken)
+        {
+            RunService(token =>
+            {
+                var solution = SolutionService.PrimaryWorkspace.CurrentSolution;
+                var documentId = GetDocumentId(solution, projectIdGuid, filePath);
+                if (documentId == null)
+                {
+                    return;
+                }
+
+                SemanticChangeTracker.Track(solution.Workspace, Rpc, documentId);
+            }, cancellationToken);
+        }
+
+        private static DocumentId GetDocumentId(Solution solution, Guid projectIdGuid, string filePath)
+        {
+            var documentIds = solution.GetDocumentIdsWithFilePath(filePath);
+
+            if (projectIdGuid == Guid.Empty)
+            {
+                // this is misc project case. in this case, just return first match if there is one
+                return documentIds.FirstOrDefault();
+            }
+
+            var projectId = ProjectId.CreateFromSerialized(projectIdGuid);
+            return documentIds.FirstOrDefault(id => id.ProjectId == projectId);
+        }
+
+        private class SemanticChangeTracker
+        {
+            private const int DelayInMS = 100;
+
+            private readonly Workspace _workspace;
+            private readonly JsonRpc _rpc;
+            private readonly DocumentId _documentId;
+
+            private readonly object _gate;
+
+            private ResettableDelay _resettableDelay;
+
+            public static void Track(Workspace workspace, JsonRpc rpc, DocumentId documentId)
+            {
+                var _ = new SemanticChangeTracker(workspace, rpc, documentId);
+            }
+
+            private SemanticChangeTracker(Workspace workspace, JsonRpc rpc, DocumentId documentId)
+            {
+                _gate = new object();
+
+                _workspace = workspace;
+                _rpc = rpc;
+
+                _documentId = documentId;
+
+                _resettableDelay = ResettableDelay.CompletedDelay;
+
+                ConnectEvents(subscription: true);
+            }
+
+            private void ConnectEvents(bool subscription)
+            {
+                lock (_gate)
+                {
+                    if (subscription)
+                    {
+                        _rpc.Disconnected += OnRpcDisconnected;
+                        _workspace.WorkspaceChanged += OnWorkspaceChanged;
+                    }
+                    else
+                    {
+                        _rpc.Disconnected -= OnRpcDisconnected;
+                        _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+                    }
+                }
```

Comment:
it feels weird to have one method that does this with a bool... instead of just a Connect/Disconnect pair...

Reply:
wanted to have 1 place where lock is used.